---
title: "Notes on Bis-SNP"
author: "Peter Hickey"
date: "31 October 2014"
output: html_document
---

Some notes on `Bis-SNP` output files using EPISCOPE E13BUF data.

```{r}
library(VariantAnnotation)
x <- readVcf(file = "/Volumes/hickey/Adipose/E13BUF/E13BUF.realign.recal.snp.filtered.sort.vcf.gz", genome = 'hg19')
x
```

Look at the `rowRanges`:

```{r}
rowRanges(x)
table(strand(x))
table(rowRanges(x)$REF)
table(rowRanges(x)$FILTER, useNA = "ifany")
```

Look at the `info`:

```{r}
info(x)
table(info(x)$CS, useNA = "ifany")
# What's the REF at positions where the strand is NA
table(as.character(rowRanges(x)$REF[is.na(info(x)$CS)]))
table(info(x)$Context@unlistData, useNA = "ifany")
table(info(x)$DB, useNA = "ifany")
table(info(x)$DB, useNA = "ifany") / nrow(x) * 100
plot(density(info(x)$DP))
table(info(x)$HQ, useNA = "ifany")
table(info(x)$MQ0, useNA = "ifany")
table(info(x)$NS, useNA = "ifany")
table(info(x)$QD, useNA = "ifany")
table(info(x)$REF@unlistData, useNA = "ifany")
plot(density(info(x)$SB, na.rm = TRUE))
```